beacon_node/beacon_chain/src/light_client_finality_update_verification.rs

```diff
@@ -0,0 +1,94 @@
+/// - The application encountered an internal error whilst attempting to determine validity
+///   (the `BeaconChainError` variant)
+#[derive(Debug, AsRefStr)]
+pub enum Error {
+    /// Light client finality update message with a lower or equal finalized_header slot already forwarded.
+    FinalityUpdateAlreadySeen,
+    /// The light client finality message was received is prior to one-third of slot duration passage. (with
+    /// respect to the gossip clock disparity and slot clock duration).
+    ///
+    /// ## Peer scoring
+    ///
+    /// Assuming the local clock is correct, the peer has sent an invalid message.
+    TooEarly,
+    /// Light client finality update message does not match the locally constructed one.
+    InvalidLightClientFinalityUpdate,
+    // Signature slot start time is none.
+    SigSlotStartIsNone,
+    // Beacon chain error occured.
+    BeaconChainError(BeaconChainError),
+}
+
+impl From<BeaconChainError> for Error {
+    fn from(e: BeaconChainError) -> Self {
+        Error::BeaconChainError(e)
+    }
+}
+
+/// Wraps a `LightClientFinalityUpdate` that has been verified for propagation on the gossip network.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "T: BeaconChainTypes"))]
+pub struct VerifiedLightClientFinalityUpdate<T: BeaconChainTypes> {
+    light_client_finality_update: LightClientFinalityUpdate<T::EthSpec>,
+    seen_timestamp: Duration,
+}
+
+impl<T: BeaconChainTypes> VerifiedLightClientFinalityUpdate<T> {
+    /// Returns `Ok(Self)` if the `light_client_finality_update` is valid to be (re)published on the gossip
+    /// network.
+    pub fn verify(
+        light_client_finality_update: LightClientFinalityUpdate<T::EthSpec>,
+        chain: &BeaconChain<T>,
+        seen_timestamp: Duration,
+    ) -> Result<Self, Error> {
+        let gossiped_finality_slot = light_client_finality_update.finalized_header.slot;
+        let local_finality_update_slot = chain.get_finalized_slot()?;
+        let one_third_slot_duration = Duration::new(chain.spec.seconds_per_slot / 3, 0);
+        let signature_slot = light_client_finality_update.signature_slot;
+
+        let start_time = chain.slot_clock.start_of(signature_slot);
+
+        // verify that no other finality_update with a lower or equal
+        // finalized_header.slot was already forwarded on the network
+        if local_finality_update_slot >= gossiped_finality_slot {
+            return Err(Error::FinalityUpdateAlreadySeen);
+        }
+
+        // verify that enough time has passed for the block to have been propagated
+        match start_time {
+            Some(time) => {
+                if seen_timestamp + MAXIMUM_GOSSIP_CLOCK_DISPARITY > time + one_third_slot_duration
+                {
+                    return Err(Error::TooEarly);
+                }
+            }
+            None => return Err(Error::SigSlotStartIsNone),
+        }
+
+        // TODO: fetch locally contstructed LightClientFinalityUpdate from cache
```

Comment:
I haven't thought much about how to structure the cache yet. It's somewhat involved because we need to serve RPC requests for the last ~5 months of chain history.

Reply:
Thank you for reviewing this stuff! In the spec the last validation check for full nodes is: 

"[IGNORE] The received finality_update matches the locally computed one exactly (as defined in create_light_client_finality_update".

With the cache I figured what I had commented in that line would work fine. Prior to the cache being implemented should I  create_light_client_finality_update on each verification to test? That doesn’t seem great right?

Reply:
Actually I think if we have just the `latest_seen_update` cached then this check won't be too expensive, we'll compute the latest finality update just once when we first receive it on gossip, then ignore all future updates

Reply:
That makes sense. I will fix it then set it ready for review. Once the finality topic is done I'll immediately do the optimistic one.

Reply:
I think you still need to compute it with the proof methods added in #3674. At the moment the update is just compared to the default and rejected.

To get a mutable `BeaconState` will be a bit annoying, you'll have to `.clone()` the head snapshot. This is fine for prototyping, but before we merge and enable this we should probably re-work the proof methods to _assume_ a built tree hash cache and just take it by `&` reference. I can have a quick go at this tomorrow

Reply:
Ok I had a go at adding a suitable method to the `BeaconChain` here: https://github.com/sigp/lighthouse/pull/3714

Usage here would be something like:

```rust
let head = chain.head_snapshot();
chain.with_mutable_state_for_block(&head.beacon_block, head.beacon_block_root, |state, cache_hit| {
    state.initialize_tree_hash_cache();
    
    // Log some stuff about `cache_hit` at debug level
    // ....
    
    // Build update with proof
    Ok(LightClientFinalityUpdate::from_state(state))
})
```

Reply:
Sorry if this is a dumb question, but can I make this branch dependent on yours?

Reply:
Yeah you can rebase your changes on mine by doing something like:

```
git remote add michael https://github.com/michaelsproul/lighthouse.git

git fetch michael

git rebase -i michael/light-client-groundwork
```

Another option is merging my branch to yours rather than rebasing, although that's usually only necessary if there are lots of commits to resolve conflicts between

Reply:
I'm really unsure how to implement this method: LightClientFinalityUpdate::from_state(state). I see the proof methods being used for LightClientFinalityUpdate::new(), but to use that method I'm having trouble accessing the finalized block needed,

Reply:
I think things are working now. I think it could be made cleaner by some of the type magic you speak of but  for now I was just trying to make it work. So with super struct are you suggesting making LightClientFinalityUpdate a superstruct with variants(Base, Signed) or something? If so, when the gossiped_update is compared with the locally constructed one wouldn't equality always fail? (assuming the gossiped one would be the Base variant and locally constructed one would be Signed) Or I probably just don't understand the idea.

Reply:
Nah, no need to make the `LightClientUpdate` a `superstruct`. I was talking about using `BeaconBlockRef` which is a type generated by `superstruct` for `BeaconBlock`, but it's not necessary.

Lets continue the discussion here: https://github.com/sigp/lighthouse/pull/3693#discussion_r1024611291